core/pta/widevine.c

```diff
@@ -0,0 +1,188 @@
+ * The format of data should be TPM2B_PUBLIC.
+ * For more information, please reference the 12.2.5 section:
+ * https://trustedcomputinggroup.org/wp-content/uploads/TCG_TPM2_r1p59_Part2_Structures_pub.pdf
+ */
+static uint8_t tpm_auth_pub[TPM_AUTH_PUB_MAX_SIZE];
+static uint32_t tpm_auth_pub_size;
+
+/*
+ * The Widevine root of trust secret. Used to sign the widevine
+ * request in OP-TEE. The value is an ECC NIST P-256 scalar.
+ * For more information, please reference the G.1.2 section:
+ * https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-186.pdf
+ */
+static uint8_t widevine_priv[WIDEVINE_PRIV_MAX_SIZE];
+static uint32_t widevine_priv_size;
+
+static TEE_Result init_widevine_dt_data(void)
+{
+	int node = 0;
+	int len = 0;
+	void *fdt = NULL;
+	const void *value = NULL;
+
+	fdt = get_secure_dt();
+	if (!fdt)
+		return TEE_ERROR_NO_DATA;
+
+	node = fdt_path_offset(fdt, "/options/widevine");
+	if (node < 0)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	value = fdt_getprop(fdt, node, "tcg-tpm-auth-public-key", &len);
+	if (!value)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	if (len > TPM_AUTH_PUB_MAX_SIZE)
+		return TEE_ERROR_OVERFLOW;
+
+	memcpy(tpm_auth_pub, value, len);
+	tpm_auth_pub_size = len;
+
+	value = fdt_getprop(fdt, node, "widevine-root-of-trust", &len);
+	if (!value)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	if (len > WIDEVINE_PRIV_MAX_SIZE)
+		return TEE_ERROR_OVERFLOW;
+
+	memcpy(widevine_priv, value, len);
+	widevine_priv_size = len;
+
+	return TEE_SUCCESS;
+}
+
+service_init(init_widevine_dt_data);
+
+static TEE_Result open_session(uint32_t param_types __unused,
+			       TEE_Param params[TEE_NUM_PARAMS] __unused,
+			       void **sess_ctx __unused)
+{
+	size_t i;
+	struct ts_session *session = ts_get_calling_session();
+
+	/* Make sure we called from Secure World */
+	if (!session)
+		return TEE_ERROR_ACCESS_DENIED;
+
+	/* Make sure we called from a TA */
+	if (to_ta_session(ts_get_current_session())->clnt_id.login !=
+	    TEE_LOGIN_TRUSTED_APP)
+		return TEE_ERROR_ACCESS_DENIED;
+
+	/* Make sure we called from an allowed TA */
+	for (i = 0; i < ARRAY_SIZE(allowed_ta_uuids); i++)
+		if (memcmp(&session->ctx->uuid, &allowed_ta_uuids[i],
+			   sizeof(TEE_UUID)) == 0)
+			return TEE_SUCCESS;
+
+	return TEE_ERROR_ACCESS_DENIED;
+}
+
+static TEE_Result get_dt_data(uint32_t ptypes, TEE_Param params[TEE_NUM_PARAMS],
+			      uint32_t cmd)
+{
+	uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+					  TEE_PARAM_TYPE_NONE,
+					  TEE_PARAM_TYPE_NONE,
+					  TEE_PARAM_TYPE_NONE);
+
+	uint8_t *data = NULL;
+	uint32_t data_length = 0;
+
+	if (exp_pt != ptypes)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (cmd == PTA_WIDEVINE_GET_TPM_PUBKEY) {
+		data = tpm_auth_pub;
+		data_length = tpm_auth_pub_size;
+	} else if (cmd == PTA_WIDEVINE_GET_WIDEVINE_PRIVKEY) {
+		data = widevine_priv;
+		data_length = widevine_priv_size;
+	} else {
+		return TEE_ERROR_NOT_IMPLEMENTED;
+	}
+
+	params[0].memref.size = data_length;
```

Comment:
Must test `data_length` against `params[0].memref.size`:
```c
	if (data_length < params[0].memref.size) {
		params[0].memref.size = data_length;
		return TEE_ERROR_SHORT_BUFFER;
	}
```

I think you should return an error if there `data_length` is zero. Unless you're fine to return `TEE_SUCCESS` and let client consider zero sized buffer means the requested key was not provisioned.